src/Microsoft.Extensions.DotnetToolDispatcher.Sources/DotnetToolDispatcher.cs

```diff
@@ -0,0 +1,98 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Reflection;
+using Microsoft.DotNet.Cli.Utils;
+using Microsoft.DotNet.ProjectModel;
+using Microsoft.Extensions.CommandLineUtils;
+using Microsoft.Extensions.PlatformAbstractions;
+
+namespace Microsoft.Extensions.Internal
+{
+    internal static class DotnetToolDispatcher
+    {
+        private const string DispatcherVersionArgumentName = "--dispatcher-version";
+
+        private static readonly string ToolName = PlatformServices.Default.Application.ApplicationName;
+        private static readonly string DispatcherVersionArgumentValue = ResolveDispatcherVersionArgumentValue();
+
+        public static ICommand CreateDispatchCommand(
+            IEnumerable<string> dispatchArgs,
+            ProjectContext projectContext,
+            string configuration,
+            string outputPath,
+            string buildBasePath)
+        {
+            configuration = configuration ?? Constants.DefaultConfiguration;
+            var commandFactory = new ProjectDependenciesCommandFactory(
+                projectContext.TargetFramework,
+                configuration,
+                outputPath,
+                buildBasePath,
+                projectContext.ProjectDirectory);
+
+            var dispatchArgsList = new List<string>(dispatchArgs);
+            dispatchArgsList.Add(DispatcherVersionArgumentName);
+            dispatchArgsList.Add(DispatcherVersionArgumentValue);
+
+            var command = commandFactory.Create(
+                ToolName,
+                dispatchArgsList,
+                projectContext.TargetFramework,
+                configuration);
+            return command;
+        }
+
+        public static bool IsDispatcher(string[] programArgs) =>
+            !programArgs.Contains(DispatcherVersionArgumentName, StringComparer.OrdinalIgnoreCase);
+
+        public static void EnsureValidDispatchRecipient(CommandLineApplication application, string[] programArgs)
+        {
+            Debug.Assert(!IsDispatcher(programArgs));
+
+            if (!application.Options.Any(option =>
+                string.Equals(option.Template, DispatcherVersionArgumentName, StringComparison.OrdinalIgnoreCase)))
+            {
+                // Add the dispatch argument to ensure no errors are caused by unhandled arguments.
+                application.Option(
```

Comment:
Yeah, we do something similar here: https://github.com/dotnet/cli/blob/1a8d393b9be8f6f30d4fd6c227c5bdb343c8c8e1/src/Microsoft.DotNet.Cli.Utils/DebugHelper.cs#L13-L20